### ISSUE TITLE
Unable to explicitly exclude only directory name in root

### ISSUE THREAD

**Issue by Goblinlordx** [Microsoft/vscode#42174]
- VSCode Version: 1.19.3
- OS Version: OSX 10.13.2

Steps to Reproduce:

1. I have a project structure in the form of:
```
- path1/
  - file_to_exclude_1
  - file_to_exclude_2
  - file_to_exclude_3
- path2/
  - file_to_include_1
  - path1/
    - file_to_include_2
```
2. In either `User Settings` or `Workspace Settings` I have:
```
"files.exclude": {
        "path1/*": true
}
```

**Expected result:** 
1. `./path1/` and it's contents are excluded
  - contents are not shown in the file tree
  - search does not contain results from those files. 
2. `./path2/path1/` and it's contents not excluded
  - displays in file tree
  - shows search results within the path.

**Actual result:** 
1. `./path1/` and it's contents are excluded
  - contents are not shown in the file tree
  - search does not contain results from those files. 
2. `./path2/path1/` and it's contents are **partially** excluded
  - contents displays in file tree
  - **search does not contain results from those files**

I would expect that all files which are shown in the file tree to be searched (with the exception of anything excluded via the "files.exclude").  This, however, does not seem to be the current behavior.  This is also pretty confusing because when a file inside `/path2/path1/` is "open" in a tab then it will still show search results for that file (when using `Find in Files`.

Based on usage of glob patterns I kind of expected `/path1/` would be excluded completely but thought it was working based on only `/path1/` being removed from the file tree display.

Does this issue occur when all extensions are disabled?: Yes

**Comment by roblourens** [comment]
> /path2/path1/ and it's contents are partially excluded
> search does not contain results from those files

I don't see this behavior - I set up this scenario and search does show results from those files. Do you see that file in quickopen? (cmd+p)? 

Can you show the output in the search output pane?

![image](https://user-images.githubusercontent.com/323878/32013959-89edf3e4-b971-11e7-8f3d-9f4b4b7790e9.png)

**Comment by Goblinlordx** [comment]
Sure.  There are multiple source file types in this project so to get this though the ones I am looking for don't show up when I restrict it to just `*.js`.  In this case, I did restrict it and got:
```
rg --hidden --heading --line-number --color 'ansi' --colors 'path:none' --colors 'line:none' --colors 'match:fg:red' --colors 'match:style:nobold' --ignore-case -g '/Users/benbaldivia/dev/buzzscreen/**/*.js/**' -g '/Users/benbaldivia/dev/buzzscreen/**/*.js' -g '!/Users/benbaldivia/dev/buzzscreen/**/.git' -g '!/Users/benbaldivia/dev/buzzscreen/**/.svn' -g '!/Users/benbaldivia/dev/buzzscreen/**/.hg' -g '!/Users/benbaldivia/dev/buzzscreen/**/CVS' -g '!/Users/benbaldivia/dev/buzzscreen/**/.DS_Store' -g '!/Users/benbaldivia/dev/buzzscreen/static/*' -g '!/Users/benbaldivia/dev/buzzscreen/**/node_modules' -g '!/Users/benbaldivia/dev/buzzscreen/**/bower_components' -g '!/Users/benbaldivia/dev/buzzscreen/**/vendor' --max-filesize '17179869184' --follow --fixed-strings -- 'ACCESS' '/Users/benbaldivia/dev/buzzscreen'
 - cwd: /
 - Sibling clauses: {}
No files were searched, which means ripgrep probably applied a filter you didn't expect. Try running again with --debug.
```

When I do it with one file open I get:
![image](https://user-images.githubusercontent.com/2612830/35425979-d826ca34-02a2-11e8-9355-c687ac0f2a7f.png)
```
rg --hidden --heading --line-number --color 'ansi' --colors 'path:none' --colors 'line:none' --colors 'match:fg:red' --colors 'match:style:nobold' --ignore-case -g '/Users/benbaldivia/dev/buzzscreen/**/*.js/**' -g '/Users/benbaldivia/dev/buzzscreen/**/*.js' -g '!/Users/benbaldivia/dev/buzzscreen/**/.git' -g '!/Users/benbaldivia/dev/buzzscreen/**/.svn' -g '!/Users/benbaldivia/dev/buzzscreen/**/.hg' -g '!/Users/benbaldivia/dev/buzzscreen/**/CVS' -g '!/Users/benbaldivia/dev/buzzscreen/**/.DS_Store' -g '!/Users/benbaldivia/dev/buzzscreen/static/*' -g '!/Users/benbaldivia/dev/buzzscreen/**/node_modules' -g '!/Users/benbaldivia/dev/buzzscreen/**/bower_components' -g '!/Users/benbaldivia/dev/buzzscreen/**/vendor' --max-filesize '17179869184' --follow --fixed-strings -- 'ACCESS' '/Users/benbaldivia/dev/buzzscreen'
 - cwd: /
 - Sibling clauses: {}
No files were searched, which means ripgrep probably applied a filter you didn't expect. Try running again with --debug.
```

If I specifically include the folder name which I added to `files.exclude` then I get the type of results I was expecting:
![image](https://user-images.githubusercontent.com/2612830/35426033-5925b208-02a3-11e8-8256-32e4b7a8bfa8.png)
```
rg --hidden --heading --line-number --color 'ansi' --colors 'path:none' --colors 'line:none' --colors 'match:fg:red' --colors 'match:style:nobold' --ignore-case -g '/Users/benbaldivia/dev/buzzscreen/**/static/**' -g '/Users/benbaldivia/dev/buzzscreen/**/static' -g '!/Users/benbaldivia/dev/buzzscreen/**/.git' -g '!/Users/benbaldivia/dev/buzzscreen/**/.svn' -g '!/Users/benbaldivia/dev/buzzscreen/**/.hg' -g '!/Users/benbaldivia/dev/buzzscreen/**/CVS' -g '!/Users/benbaldivia/dev/buzzscreen/**/.DS_Store' -g '!/Users/benbaldivia/dev/buzzscreen/static/*' -g '!/Users/benbaldivia/dev/buzzscreen/**/node_modules' -g '!/Users/benbaldivia/dev/buzzscreen/**/bower_components' -g '!/Users/benbaldivia/dev/buzzscreen/**/vendor' --max-filesize '17179869184' --follow --fixed-strings -- 'ACCESS' '/Users/benbaldivia/dev/buzzscreen'
 - cwd: /
 - Sibling clauses: {}
```

**Comment by Goblinlordx** [comment]
I forgot to answer your initial question.  I see files that I want to search in the `recently opened` area of the quick open before searching and after typing.
![image](https://user-images.githubusercontent.com/2612830/35426478-c444e10a-02a6-11e8-8fdb-9486101cac67.png)

**Comment by Goblinlordx** [comment]
I noticed `ripgrep` does ignore `.gitignore` files.  My `.gitignore` does have an entry the same as I have in `files.exclude`:
```
static/*
```

Even so, it doesn't make much sense that without the `exclude` it would search `./static/*` and not search `./ui/static/*`.  Without the `files.exclude` entry, the folder I don't want to see does display in the file tree.

Another note, when I have `files.exclude` set I can include the same thing and get what I am looking as mentioned above.  I am currently doing this as a workaround but doing it this way doesn't allow me to restrict it to a filetype or something like that.  When using `static/**/*.js` it actually results in nothing (except in files that are open) which is the same as if I don't explicitly include it.

**Comment by roblourens** [comment]
ripgrep does look at your .gitignore file, and `static/*` should match any folder `static` in any subfolder. `/static/*` I think would root it to the top level folder. Does that basically explain your problem? If you want to just disable that behavior for ripgrep, you can set `"search.useIgnoreFiles": false` and see if that does it for you.

I think that ripgrep will let includes override .gitignore patterns, but only if they match exactly the same thing, so you may be able to cancel out the gitignore pattern with `static/*` or `static`, but not `static/**/*.js`.

**Comment by Goblinlordx** [comment]
That would also mean ripgrep is using `.gitignore` differently than git though.  Currently, git is ignoring only the folder in root but not the path in subdirectories.

**Comment by Goblinlordx** [comment]
Changing `static/*` in .gitignore to `/static/*` fixed this issue.  As for the difference in behavior between what `git` actually ignores when using `.gitignore` and what `ripgrep` ignores, I think this is an issue with `ripgrep`.  Thanks for looking into this and pointing me in the correct direction.

**Comment by roblourens** [comment]
Yeah, testing this myself, there's a difference between `static/*` and `static/` in a gitignore file, which I didn't realize.

**Comment by roblourens** [comment]
Thanks for following through and filing that upstream issue. If the PR is accepted I'll update the version of ripgrep that we use.

**Comment by Goblinlordx** [comment]
@roblourens 
Upstream fix has been merged and is included in newest release:
https://github.com/BurntSushi/ripgrep/releases/tag/0.8.0

**Comment by Goblinlordx** [comment]
@robertohuertasm Any update on this?  There seem to have been a 2 releases since but it seems the update still hasn't been pulled in.

**Comment by roblourens** [comment]
This is fixed in 1.22, insiders or stable release coming next week. The last ripgrep update was delayed from 1.21 due to some other issues.

**Comment by Goblinlordx** [comment]
:+1: Thanks~
